### ISSUE TITLE
feat(beam): diverse beam search with group-sequential factorization and OpenAI API gateway parameters

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12413]
# feat(beam): diverse beam search with group-sequential factorization and OpenAI API gateway parameters

```routing
lane: internal/beam
paths: internal/beam/diverse.go, internal/gateway/chat_params.go, internal/beam/diverse_test.go, internal/gateway/beam_test.go
expected_steps: 4
```

**Lane:** internal/beam  
**Work Unit:** leaf  
**Generation:** gen/now  
**Priority:** P2  

## Parent context
Extends #12407 (`feat(beam): zero-allocation beam search engine spine`) and #12411 (`feat(beam): path-dependent grammar and schema-constrained beam search`). Exposes beam search parameters through the OpenAI-compatible gateway API and implements Diverse Beam Search (Vijayakumar et al.) to prevent beam collapse.

## Current state
Standard beam search suffers from search collapse: surviving hypotheses often share identical prefixes and differ only by minor punctuation or inflectional suffixes. Furthermore, incoming OpenAI-compatible Chat Completion requests (`/v1/chat/completions`) in `internal/gateway` currently discard or reject beam search parameters (`num_beams`, `diversity_penalty`, `length_penalty`, `no_repeat_ngram_size`).

## Why this is next
Diverse Beam Search partitions beams into groups and penalizes token overlap across groups, forcing the engine to explore genuinely distinct modes in the probability distribution. Exposing these parameters through `internal/gateway` allows standard OpenAI API clients and agent harnesses to request deterministic, diverse beam search without modifying upstream client SDKs.

## Working spine
Partition $B$ beams into $G$ groups -> decode groups sequentially with Hamming diversity penalties against prior groups -> enforce $N$-gram blocking to kill cyclic loops -> expose `num_beams`, `diversity_penalty`, `length_penalty`, and `no_repeat_ngram_size` in `/v1/chat/completions` -> verify diverse output hypotheses and gateway conformance.

## Core through-line
1. Implement Diverse Beam Search in `internal/beam/diverse.go`: partition $B$ beams into $G$ groups, decoding sequentially at step $t$ and applying a token Hamming diversity penalty:
   $$\tilde{S}(y_t \mid y_{<t}^{(b)}; g) = S(y_{<t}^{(b)}) + \log P(y_t \mid y_{<t}^{(b)}) - \lambda \sum_{k=1}^{g-1} \mathbb{I}[y_t \in \mathbf{Y}_t^{(k)}]$$
2. Implement dynamic $N$-gram repetition blocking (`no_repeat_ngram_size`): track rolling $N$-gram hashes per beam and set logits of repeating tokens to $-\infty$.
3. Expose beam search parameters in `internal/gateway/chat_params.go` and `pkg/abi`:
   - `num_beams` (int, default: 1 = greedy/sampling)
   - `length_penalty` (float32, default: 1.0)
   - `diversity_penalty` (float32, default: 0.0)
   - `no_repeat_ngram_size` (int, default: 0 = disabled)
   - `early_stopping` (bool | string, default: true)
4. Author end-to-end gateway integration tests in `internal/gateway/beam_test.go` asserting that diverse beam outputs contain distinct semantic completions and respect $N$-gram blocking.

## In scope
- Diverse Beam Search with group-sequential factorization.
- Dynamic $N$-gram blocking to prevent repetitive output loops.
- Parameter ingestion in `internal/gateway` via standard and `extra_body` payloads.
- Integration tests witnessing diverse outputs and gateway contract conformance.

## Out of scope
- Continuous batching multi-tenant multiplexing (beam requests run as isolated slot sessions).
- External vector-embedding distance diversity (Hamming token diversity is sufficient and zero-alloc).

## Gold-plating boundary
Strictly excluded:
- Complex clustering heuristics beyond group-sequential Hamming diversity.
- Deprecated legacy completion endpoints.

## Verifiable Witness
```bash
go test -v ./internal/beam -run TestDiverseBeamSearch
go test -v ./internal/gateway -run TestGatewayBeamSearchParameters
```

## Done condition
Gateway endpoint parses beam search parameters, invokes `internal/beam` with Diverse Beam Search, enforces $N$-gram blocking, and returns diverse, high-likelihood completions across surviving beams.

## Scoped Acceptance Criteria
- [ ] 1. Diverse Beam Search produces distinct completions across groups with zero identical prefixes. `[SW-VERIFIED]`
- [ ] 2. `no_repeat_ngram_size` strictly blocks all cyclic repetition patterns. `[SW-VERIFIED]`
- [ ] 3. `/v1/chat/completions` correctly parses `num_beams`, `diversity_penalty`, and `length_penalty`. `[SW-VERIFIED]`
- [ ] 4. Gateway responses return the winning beam or all $B$ candidate choices when requested. `[SW-VERIFIED]`

## Acceptance gate
`go test -v ./internal/beam -run TestDiverseBeam` and `go test -v ./internal/gateway -run TestGatewayBeam` exit 0 with 0 failures.

## Closure binding
Close via an explicit-path trunk commit landing `internal/beam/diverse.go` and `internal/gateway/` parameter bindings.

## Likely files
- `internal/beam/diverse.go`
- `internal/gateway/chat_params.go`
- `internal/beam/diverse_test.go`
- `internal/gateway/beam_test.go`
